src/compiler/counted_array.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 Mozilla Foundation
+// Original Copyright (c) 2016 Alex Burka, licensed to Mozilla Foundation under MIT license
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Helper macro to create static array of calculable fixed length
+#[macro_export]
+macro_rules! counted_array {
+    // INTERNAL
+
+    // last element (proceed to output)
+    (@parse $size:expr, ($val:expr) -> [$($accs:expr),*] $thru:tt) => {
+        counted_array!(@output $size + 1usize, [$($accs,)* $val] $thru);
+    };
+    // more elements (keep parsing)
+    (@parse $size:expr, ($val:expr, $($vals:expr),*) -> [$($accs:expr),*] $thru:tt) => {
+        counted_array!(@parse $size + 1usize, ($($vals),*) -> [$($accs,)* $val] $thru);
+    };
+
+    // output a local variable
+    (@output $size:expr, $acc:tt (() let $n:ident $t:ty)) => {
+        let $n: [$t; $size] = $acc;
+    };
+    // output a static or const item
+    (@output $size:expr, $acc:tt (($($p:tt)*) $s:ident $n:ident $t:ty)) => {
+        $($p)* $s $n: [$t; $size] = $acc;
+    };
+
+    // EXTERNAL
+
+    // entry point
+    (pub $storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+        counted_array!(@parse 0usize, ($($vals),*) -> [] ((pub) $storage $n $t));
+    };
+    (pub $restr:tt $storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+        counted_array!(@parse 0usize, ($($vals),*) -> [] ((pub $restr) $storage $n $t));
+    };
+    ($storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+        counted_array!(@parse 0usize, ($($vals),*) -> [] (() $storage $n $t));
+    };
+}
+
+#[cfg(test)]
+mod counted_array_test {
+
+    #[test]
+    fn counted_array_macro() {
+        counted_array!(static ARR_QUAD: [u8;_] = [1,2,3,4,]);
+        assert_eq!(ARR_QUAD.len(), 4);
+    }
+
+    #[test]
+    fn private_arrays() {
+        counted_array!(const LOCAL_CONST_ARR: [i32; _] = [1, 2, 3]);
+        counted_array!(let local_arr: [i32; _] = [4, 5, 6]);
```

Comment:
We only ever use `static ARR` variants, everything else is dead weight we have to maintain.

Reply:
Fixed.